### DISCUSSION TITLE
NextJS - Maintain cache while navigating between different pages

### DISCUSSION THREAD

**Discussion by george-roussos** [TanStack/query#7518] [Q&A]
Hi, this is my first time using react-query on a NextJS project.

Using it in a single page React app has been no problems re: the shared cache. Like, you go in one page, navigate to another, go back to the first page, data is cached, served, done.

I am realising just now that with NextJS that is not really the case, since the docs advise on every page having its own `queryClient` instance. This is good and motivated in the docs, however, I was hoping to benefit from the cache while the user is navigating between different pages. Is there a way to do this? I am guessing with a shared, exported `queryClient` instance this might be possible - but what are the caveats and what is the correct way to do so accounting for hydration (right now I have the hydration boundary wrapping everything in my `_app.tsx` (using `pages/` structure). I don't believe my pages expose any sensitive data (I get my data from a contentful instance and no authentication or similar is involved).

By the way, thanks for this tool, what a thing...

**Comment by TkDodo** [comment]
> I am realising just now that with NextJS that is not really the case, since the docs advise on every page having its own queryClient instance.

That's not true. You should have one `QueryClient` per browser session - either by sharing it over the `window` or by instantiating it in React State inside the `Providers` component, where you'll only have one per application.

**Reply by george-roussos** [reply]
Ok, then it might be worth to mention this in the docs, because looking at the guide here (specifically for `NextJS`):

https://tanstack.com/query/latest/docs/framework/react/guides/ssr

In each route (which has its own `getStaticProps()` definition) a new `queryClient` instance is created and the queries are prefetched using that client. For someone that has never used the library with SSR before this could be mistaken as advice for every page to have its own client instance.

But! Thank you very much for clearing it up and answering my question.

**Reply by TkDodo** [reply]
the code you are pointing to is for the server side part - where data is fetched on the server. This indeed is not cached and everyone gets their own `queryClient`.

But this is _not_ the QueryClient that is used in the Browser. That one we are creating in `_app.tsx`, and there, you can see that we have it inside React State, so it's only one per client lifecycle:

```
// _app.tsx
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

// NEVER DO THIS:
// const queryClient = new QueryClient()
//
// Creating the queryClient at the file root level makes the cache shared
// between all requests and means _all_ data gets passed to _all_ users.
// Besides being bad for performance, this also leaks any sensitive data.

export default function MyApp({ Component, pageProps }) {
  // Instead do this, which ensures each request has its own cache:
  const [queryClient] = React.useState(
    () =>
      new QueryClient({
        defaultOptions: {
          queries: {
            // With SSR, we usually want to set some default staleTime
            // above 0 to avoid refetching immediately on the client
            staleTime: 60 * 1000,
          },
        },
      }),
  )

  return (
    <QueryClientProvider client={queryClient}>
      <Component {...pageProps} />
    </QueryClientProvider>
  )
}
```

**Reply by george-roussos** [reply]
Aha ok. Does that then mean that when querying in components, the client query is used? Which means that data between page changes is cached? Forgive me if it looks trivial but I am trying to understand how it works. My assumption was (wrong, I presume), since we are fetching server-side under a new instance, this removes any old cache. Looking at the devtools, if I navigate to a different page, the cached data for the page I was on is gone. So if page 1 -> page 2 (using NextJS's `<Link/>`) the data for page 1 is not available in the cache anymore. Maybe my implementation is wrong.

**Reply by TkDodo** [reply]
> Does that then mean that when querying in components, the client query is used?

yes

> Which means that data between page changes is cached

yes

The newly created cache on the server is empty, but we only need to send data to the client. There, the `HydrationBoundary` takes the data from the server and puts it into the client cache

**Reply by george-roussos** [reply]
Aha ok, then I think my implementation might be wrong when using `<Link/>`. Thank you for clarifying 🙏

**Reply by shamun-khatri** [reply]
@george-roussos @TkDodo I have implemented the react query in my next js 15 app router and seems like the data is not cached between routes like if I go to specific route and it fetches and cheche the data but when I go to another page and  then comeback it fetches again how do I fix  this?
I have properly set staleTime

**Reply by george-roussos** [reply]
> @george-roussos @TkDodo I have implemented the react query in my next js 15 app router and seems like the data is not cached between routes like if I go to specific route and it fetches and cheche the data but when I go to another page and then comeback it fetches again how do I fix this? I have properly set staleTime

Hi!
I can think of two possible scenarios, either you are using a custom `server.js` which causes this behaviour, or route [staleTimes](https://nextjs.org/docs/app/api-reference/config/next-config-js/staleTimes) is too short and clashes with query's `staleTime`

**Reply by TkDodo** [reply]
> but when I go to another page and then comeback it fetches again how do I fix this?

are you talking about fetching on the server or on the client ?
